File: web/src/lib/api/grant.ts
import { request, orgPath } from "./base";

export interface ResourceGrant {
  id: number;
  resource_type: string;
  resource_id: string;
  user_id: number;
  granted_by: number;
  created_at: string;
  user?: { id: number; email: string; username: string; name?: string };
  granted_by_user?: { id: number; email: string; username: string; name?: string };
}

const resourcePlural: Record<string, string> = {
  pod: "pods",
  runner: "runners",
  repository: "repositories",
};

function grantPath(resourceType: string, resourceId: string, suffix = "") {
  const plural = resourcePlural[resourceType] || `${resourceType}s`;
  return orgPath(`/${plural}/${resourceId}/grants${suffix}`);
}

export const grantApi = {
  list: (resourceType: string, resourceId: string) =>
    request<{ grants: ResourceGrant[] }>(grantPath(resourceType, resourceId)),

  grant: (resourceType: string, resourceId: string, userId: number) =>
    request<{ grant: ResourceGrant }>(grantPath(resourceType, resourceId), {
      method: "POST",
      body: { user_id: userId },
    }),

  revoke: (resourceType: string, resourceId: string, grantId: number) =>
    request<{ message: string }>(grantPath(resourceType, resourceId, `/${grantId}`), {
      method: "DELETE",
    }),
};

File: clients/core/crates/api-client/src/modules/grant.rs
use crate::ApiClient;
use crate::error::ApiError;
use agentsmesh_types::*;

fn resource_plural(resource_type: &str) -> String {
    match resource_type {
        "pod" => "pods".into(),
        "runner" => "runners".into(),
        "repository" => "repositories".into(),
        other => format!("{other}s"),
    }
}

impl ApiClient {
    pub async fn list_resource_grants(
        &self,
        resource_type: &str,
        resource_id: &str,
    ) -> Result<ResourceGrantListResponse, ApiError> {
        let plural = resource_plural(resource_type);
        let path = self.org_path(&format!("/{plural}/{resource_id}/grants"));
        self.get(&path).await
    }

    pub async fn grant_resource_access(
        &self,
        resource_type: &str,
        resource_id: &str,
        req: &CreateResourceGrantRequest,
    ) -> Result<ResourceGrantResponse, ApiError> {
        let plural = resource_plural(resource_type);
        let path = self.org_path(&format!("/{plural}/{resource_id}/grants"));
        self.post(&path, req).await
    }

    pub async fn revoke_resource_grant(
        &self,
        resource_type: &str,
        resource_id: &str,
        grant_id: i64,
    ) -> Result<EmptyResponse, ApiError> {
        let plural = resource_plural(resource_type);
        let path = self.org_path(&format!("/{plural}/{resource_id}/grants/{grant_id}"));
        self.delete(&path).await
    }
}
